ui/chat/src/ctrl.ts

```diff
@@ -97,11 +97,19 @@ export default class ChatCtrl {
   post = (text: string): boolean => {
     text = text.trim();
     if (!text) return false;
+    if (text.startsWith('<<<<')) return false;
     if (text == 'You too!' && !this.data.lines.some(l => l.u != this.data.userId)) return false;
     if (text.length > 140) {
       alert('Max length: 140 chars. ' + text.length + ' chars used.');
       return false;
     }
+
+    if (site.analysis?.study?.relay && !site.analysis.study.relay.tourShow()) {
```

Comment:
What you have is fine.

I (personally) would not use site.analysis from chat module. Instead, I'd add a generic interface to chat/src/interfaces.ts (i.e. CustomChatHandler but with a better name), or just change [ChatPlugin](https://github.com/lichess-org/lila/blob/b58a6dab716a8fb2be8a7bdc7e15252574732fec/ui/chat/src/interfaces.ts#L26)'s existing members to optional and add two or three new bits. example:
```
interface ChatPlugin {
  ...
  prepareMsg?: (msg: string) => string; // embeds whatever special shit the
  // handler wants to embed right before sending a chat
  cleanMsg?: (msg: string) => string; // strips any special embedded shit
  // shit from a chat, used when rendering
  click?: (msg: string) => void; // chat module calls this with the whole
  // chat message, handler parses and navigates to chapter/ply if it can
}
```
Then all of this code you have here can go in relay module, except for the bits that call opts.plugin?.prepareMsg (prior to sending a chat), cleanMsg (when creating the snab vnode to render one), and click (in a click handler).

There should be a bit in relayTourView where you can create the object that implements these CustomChatHandler/ChatPlugin functions. You would pass this object to the chat module initializer via opts by way of makeChat.

But if this doesn't make sense to you, or you'd rather not do it - feel free to ignore as you've already got it working.